embassy-stm32f4/src/rtc.rs

```diff
@@ -0,0 +1,290 @@
+        interrupt::free(|cs| {
+            let alarm = &self.alarms.borrow(cs)[n];
+            alarm.callback.set(Some((callback, ctx)));
+        })
+    }
+
+    fn set_alarm(&self, n: usize, alarm_timestamp: u64) {
+        self.rtc.deref().cr1.modify(|_, w| w.cen().clear_bit());
+
+        interrupt::free(|cs| {
+            (&self.alarms.borrow(cs)[n]).timestamp.set(alarm_timestamp);
+        });
+
+        self.recompute();
+        self.rtc.deref().cr1.modify(|_, w| w.cen().set_bit());
+    }
+
+    pub fn alarm0(&'static self) -> Alarm<T> {
+        Alarm { n: 0, rtc: self }
+    }
+    pub fn alarm1(&'static self) -> Alarm<T> {
+        Alarm { n: 1, rtc: self }
+    }
+    pub fn alarm2(&'static self) -> Alarm<T> {
+        Alarm { n: 2, rtc: self }
+    }
+}
+
+impl<T: Instance> embassy::time::Clock for RTC<T> {
+    fn now(&self) -> u64 {
+        self.timestamp + self.rtc.deref().cnt.read().bits() as u64
+    }
+}
+
+pub struct Alarm<T: Instance> {
+    n: usize,
+    rtc: &'static RTC<T>,
+}
+
+impl<T: Instance> embassy::time::Alarm for Alarm<T> {
+    fn set_callback(&self, callback: fn(*mut ()), ctx: *mut ()) {
+        self.rtc.set_alarm_callback(self.n, callback, ctx);
+    }
+
+    fn set(&self, timestamp: u64) {
+        self.rtc.set_alarm(self.n, timestamp);
+    }
+
+    fn clear(&self) {
+        self.rtc.set_alarm(self.n, u64::MAX);
+    }
+}
+
+mod sealed {
+    pub trait Instance {}
+
+    impl Instance for crate::pac::TIM7 {}
+    impl Instance for crate::pac::TIM2 {}
+}
+
+/// Implemented by all RTC instances.
+pub trait Instance: sealed::Instance + Sized + 'static {
+    /// The interrupt associated with this RTC instance.
+    type Interrupt: OwnedInterrupt;
+
+    fn deref(&self) -> &tim6::RegisterBlock;
+    fn ptr() -> *const tim6::RegisterBlock;
+    fn enable_clock();
+    fn ppre(clocks: Clocks) -> u8;
+    fn pclk(clocks: Clocks) -> u32;
+}
+
+impl Instance for crate::pac::TIM7 {
+    type Interrupt = interrupt::TIM7Interrupt;
+
+    fn deref(&self) -> &tim6::RegisterBlock {
+        unsafe { &(*crate::pac::TIM7::ptr()) }
+    }
+
+    fn ptr() -> *const tim6::RegisterBlock {
+        crate::pac::TIM7::ptr() as *const _
+    }
+
+    fn ppre(clocks: Clocks) -> u8 {
+        clocks.ppre1()
+    }
+
+    fn pclk(clocks: Clocks) -> u32 {
+        clocks.pclk1().0
+    }
+
+    fn enable_clock() {
+        unsafe {
+            //NOTE(unsafe) this reference will only be used for atomic writes with no side effects
+            let rcc = &(*RCC::ptr());
+            // Enable and reset the timer peripheral, it's the same bit position for both registers
+            bb::set(&rcc.apb1enr, 5);
+            bb::set(&rcc.apb1rstr, 5);
+            bb::clear(&rcc.apb1rstr, 5);
+        }
+    }
+}
+
+impl Instance for crate::pac::TIM2 {
+    type Interrupt = interrupt::TIM2Interrupt;
+
+    fn deref(&self) -> &tim6::RegisterBlock {
+        unsafe { mem::transmute(&(*crate::pac::TIM2::ptr())) }
+    }
```

Comment:
I think TIM2 and TIM6 might be incompatible.

Reply:
TIM6 is the basic timer. TIM2 is a superset of TIM6.

Reply:
It's a superset of it in features, but not in the memory layout, which it's what counts here with `transmute`.

Reply:
Okay, so I'll adjust this to return the different registers.